framework/src/userobject/ElementQualityChecker.C

```diff
@@ -0,0 +1,236 @@
+
+  // calculate and save quality metric value for current element
+  Real mv = _current_elem->quality(_m_type);
+
+  _checked_elem_num += 1;
+  _m_sum += mv;
+  if (mv > _m_max)
+    _m_max = mv;
+  else if (mv < _m_min)
+    _m_min = mv;
+
+  // check element quality metric, save ids of elements whose quality metrics exceeds the preset
+  // bounds
+  if (mv < actual_bounds.first || mv > actual_bounds.second)
+    _elem_ids.insert(_current_elem->id());
+}
+
+void
+ElementQualityChecker::threadJoin(const UserObject & uo)
+{
+  const ElementQualityChecker & eqc = static_cast<const ElementQualityChecker &>(uo);
+  _elem_ids.insert(eqc._elem_ids.begin(), eqc._elem_ids.end());
+  _bypassed_elem_type.insert(eqc._bypassed_elem_type.begin(), eqc._bypassed_elem_type.end());
+  _bypassed += eqc._bypassed;
+  _m_sum += eqc._m_sum;
+  _checked_elem_num += eqc._checked_elem_num;
+
+  if (_m_min > eqc._m_min)
+    _m_min = eqc._m_min;
+  if (_m_max < eqc._m_max)
+    _m_max = eqc._m_max;
+}
+
+void
+ElementQualityChecker::finalize()
+{
+  _communicator.min(_m_min);
+  _communicator.max(_m_max);
+  _communicator.sum(_m_sum);
+  _communicator.sum(_checked_elem_num);
+  _communicator.set_union(_elem_ids);
+  _communicator.sum(_bypassed);
+  _communicator.set_union(_bypassed_elem_type);
+
+  if (_bypassed)
+  {
+    switch (_failure_type)
+    {
+      case FailureType::WARNING:
+        mooseWarning("Provided quality metric doesn't apply to following element type: " +
+                     Moose::stringify(_bypassed_elem_type));
+        break;
+
+      case FailureType::ERROR:
+        mooseError("Provided quality metric doesn't apply to following element type: " +
+                   Moose::stringify(_bypassed_elem_type));
+        break;
+
+      default:
+        mooseError("Unknown failure type!");
+    }
+  }
+
+  _console << libMesh::Quality::name(_m_type) << " Metric values:"
+           << "\n";
+  _console << "              Minimum: " << _m_min << "\n";
+  _console << "              Maximum: " << _m_max << "\n";
+  _console << "              Average: " << _m_sum / _checked_elem_num << "\n";
+
+  if (!_elem_ids.empty())
+  {
+    switch (_failure_type)
+    {
+      case FailureType::WARNING:
+      {
+        mooseWarning("Bad element quality in terms of ",
+                     libMesh::Quality::name(_m_type),
+                     " quality metric for ",
+                     _elem_ids.size(),
+                     " elements.");
+        break;
+      }
+
+      case FailureType::ERROR:
+      {
+        mooseError("Bad element quality in terms of ",
+                   libMesh::Quality::name(_m_type),
+                   " quality metric for ",
+                   _elem_ids.size(),
+                   " elements.");
+        break;
+      }
+
+      default:
+        mooseError("Unknown failure type!");
+    }
+
+    _console << "List of failed element IDs: "
+             << "\n";
+    _console << Moose::stringify(_elem_ids) << "\n";
```

Comment:
Is it your intent that these come out after the error message? They won't if you put them here. They will for mooseWarning since that returns, but not for mooseError().

Reply:
@permcody The logic behind the summation of value of _bypassed is to check whether there is element whose quality check was bypassed across all processors. If so, the value of _bypassed will be nonzero. Otherwise its value is zero. Depends on the value of this summation, we decide whether there is need to print out the bypassed element type or not. At least this is my intension, if the current code doesn't work as what I intent.

Reply:
So what you are saying is that _bypass is actually only used as a Boolean, not a numeric value. This means it should actually be declared and used as a Boolean. We do this kind of thing in the framework all over the place.

Aggregation:
You can use `|=` or `&=` to aggregate Booleans in threadJoin routines.
You can use Communicator::max/min to aggregate Boolean values in `finalize`.

I think this approach is cleaner than using ints. Your intention is clear and the value is stable in parallel, etc.